### DISCUSSION TITLE
`types/` diretory with global `.d.ts` type declarations

### DISCUSSION THREAD

**Discussion by wvffle** [nuxt/nuxt#22221] [Ideas]
There was a similar discussion but the author just wanted to import their types: https://github.com/nuxt/nuxt/discussions/17527

### Global types in Vue
Vue supports global type files through [`globalTypeFiles`](https://github.com/vuejs/core/blob/a3dddd62059f4a7a762046c1e7f01485b96e737d/packages/compiler-sfc/src/compileScript.ts#L75-L79) script option.

As the comment in `compileScript.ts` states, `The list must be fully resolved file system paths.`. That means users would need to define paths to the declaration files themselves and glob patterns wouldn't work.

### I'd like to propose a `types/` diretory with global `.d.ts` type declarations. 
- The `*.d.ts` files from there would be automatically added to the `globalTypeFiles` list.
- Whenever a `.d.ts` file is created/removed/renamed in that directory, Nuxt handles that automatically.

#### Additional things to consider
- The generated `tsconfig.json` could include only `../types/*.d.ts` files, so that people wouldn't run into #22209 again where their IDE says everything is fine but Vue does not see the declared types. I don't know if that would have any other implications though.

I think that this would enhance the developer experience even further. Auto imports are a great feature and no 'auto imported' types is what is missing from them in my opinion.

**Comment by localusercamp** [comment]
I think you can achieve the same behaviour by adding this folder to your tsconfig.

**Reply by niranjan-talkinglands** [reply]
not working for me... can you please provide a demo?

**Comment by mmspring13** [comment]
i'm put server types files to `server/types` directory. Example:
```
// server/types/user.d.ts

import type {UserType} from "../app/shared/user";

declare module 'h3' {
  interface H3EventContext {
    user?: UserType;
  }
}

export {}
```
it is work fine and no need update file `tsconfig.json`

**Comment by TheAlexLichter** [comment]
This is supported in Nuxt 4 via `shared/types` now 👍
